Add docs for FileSystemHandle.remove()

chrisdavidmills opened the pull request:
<!-- 🙌 Thanks for contributing to MDN Web Docs. Adding details below will help us to merge your PR faster. -->

### Description

This PR adds documentation for the new `FileSystemHandle.remove()` method. Refer to my [research document](https://docs.google.com/document/d/1ymUjOxI8yL2krSrR3Mn3ZEAGhfQ1S592qQzPV89VSyg/edit#) to find out more information on exactly what has been done.

<!-- ✍️ Summarize your changes in one or two sentences -->

### Motivation

<!-- ❓ Why are you making these changes and how do they help readers? -->

### Additional details

<!-- 🔗 Link to release notes, vendor docs, bug trackers, source control, or other places providing more context -->

### Related issues and pull requests

<!-- 🔨 If this fully resolves a GitHub issue, use "Fixes #123" -->
<!-- 👉 Highlight related pull requests using "Relates to #123" -->
<!-- ❗ If another pull request should be merged first, use "**Depends on:** #123" -->


<!-- 👷‍♀️ After submitting, go to the "Checks" tab of your PR for the build status -->

chrisdavidmills committed:
Add docs for FileSystemHandle.remove()

a-sully commented on files/en-us/web/api/filesystemhandle/remove/index.md:
```diff
@@ -0,0 +1,96 @@
+- `InvalidModificationError` {{domxref("DOMException")}}
+  - : Thrown if `recursive` is set to `false` and the entry to be removed is a directory with children.
+- `NoModificationAllowedError` {{domxref("DOMException")}}
+  - : Thrown if the browser was not able to get an exclusive lock on the entry, or if the entry is the root directory of the [Origin Private File System](/en-US/docs/Web/API/File_System_Access_API#origin_private_file_system).
```
> or if the entry is the root directory of the [Origin Private File System](/en-US/docs/Web/API/File_System_Access_API#origin_private_file_system).

This was actually recently changed to not throw an error (see [spec discussion](https://github.com/whatwg/fs/pull/9#discussion_r1042546998)).

In fact, it might be worth calling out explicitly that removing on the root of the OPFS (i.e. `navigator.storage.getDirectory()`) is a convenient way to clear the OPFS :)

chrisdavidmills replied:
Ah, OK. I've removed this bit from the "Exceptions" section and added a new line in the introduction to make people aware of this. Does it read OK?

a-sully replied:
Looks great - thanks!

a-sully commented on files/en-us/web/api/filesystemhandle/remove/index.md:
```diff
@@ -0,0 +1,96 @@
+  - : Thrown if {{domxref('PermissionStatus')}} is not `granted`.
+- `NotFoundError` {{domxref("DOMException")}}
+  - : Thrown if an entry name is not found or matched.
+- {{jsxref("TypeError")}}
+  - : Thrown if the name is not a valid string or contains characters not allowed on the file system.
```
Is this relevant? Unlike with `removeEntry()` we're not passing in any parameters here

chrisdavidmills replied:
Duh, good point. You can tell I copied this from `removeEntry()`, as it was _mostly_ the same ;-)

chrisdavidmills replied:
Removed.

a-sully commented on files/en-us/web/api/filesystemhandle/remove/index.md:
```diff
@@ -0,0 +1,96 @@
+
+## Examples
+
+Our [FileSystemFileHandle.remove() demo](https://filesystemfilehandle-remove.glitch.me/) (see the [source code](https://glitch.com/edit/#!/filesystemfilehandle-remove)) is a very simple file creator app. You can enter text into the {{htmlelement("textarea")}} and press the "Save file" {{htmlelement("button")}}, and the app will open a file picker allowing you to save that text onto your local fil ...
```
> and it doesn't sync the app with the underlying file system on page reloads

Hmm I wonder if there's a better way to word this. It seems a bit misleading since the file will still exist on the underlying file system if the user doesn't `remove()` it before closing the tab

chrisdavidmills replied:
I have updated the text to make this clearer.

a-sully commented on files/en-us/web/api/filesystemhandle/remove/index.md:
```diff
@@ -0,0 +1,96 @@
+- `NotAllowedError` {{domxref("DOMException")}}
+  - : Thrown if {{domxref('PermissionStatus')}} is not `granted`.
+- `NotFoundError` {{domxref("DOMException")}}
+  - : Thrown if an entry name is not found or matched.
```
```suggestion
  - : Thrown if the entry is not found.
```
I think?

chrisdavidmills replied:
Better! Updated.

a-sully reviewed (commented):
Mostly LGTM!

chrisdavidmills committed:
making fixes for a-sully comments

chrisdavidmills committed:
Merge branch 'main' into add-filesystemhandle-remove

a-sully reviewed (commented):
LGTM 👍

chrisdavidmills committed:
Merge branch 'main' into add-filesystemhandle-remove

Elchi3 commented on files/en-us/web/api/filesystemhandle/remove/index.md:
```diff
@@ -0,0 +1,98 @@
+
+The **`remove()`** method of the {{domxref("FileSystemHandle")}} interface requests removal of the entry represented by the handle from the underlying file system.
+
+This is a convenient addition to allow you to remove a file or directory directly from its handle. Previously you would have to obtain the handle of the parent directory, then call {{domxref("FileSystemDirectoryHandle.removeEntry()")}} on that to remove it.
```
Let's just treat this as a usual API member. It will soon not be a new addition anymore.

```suggestion
The `remove()` method allows you to remove a file or directory directly from its handle. Without this method, you would have to obtain the handle of the parent directory, then call {{domxref("FileSystemDirectoryHandle.removeEntry()")}} on that to remove it.
```

Elchi3 commented on files/en-us/web/api/filesystemhandle/remove/index.md:
```diff
@@ -0,0 +1,98 @@
+
+This is a convenient addition to allow you to remove a file or directory directly from its handle. Previously you would have to obtain the handle of the parent directory, then call {{domxref("FileSystemDirectoryHandle.removeEntry()")}} on that to remove it.
+
+You can also call `remove()` on the root directory of the [Origin Private File System](/en-US/docs/Web/API/File_System_Access_API#origin_private_file_system) as a convenient way to clear its contents, after which a new empty OPFS is created.
```
Maybe I'm grumpy but I think we don't need the advertisement :)

```suggestion
You can also call `remove()` on the root directory of the [Origin Private File System](/en-US/docs/Web/API/File_System_Access_API#origin_private_file_system) to clear its contents, after which a new empty OPFS is created.
```

Elchi3 commented on files/en-us/web/api/filesystemhandle/remove/index.md:
```diff
@@ -0,0 +1,98 @@
+
+## Examples
+
+Our [FileSystemHandle.remove() demo](https://filesystemhandle-remove.glitch.me/) (see the [source code](https://glitch.com/edit/#!/filesystemhandle-remove)) is a very simple file creator app. You can enter text into the {{htmlelement("textarea")}} and press the "Save file" {{htmlelement("button")}}, and the app will open a file picker allowing you to save that text onto your local file system in  ...
```
Nothing is ever simple. Remove for inclusivity. 

```suggestion
Our [FileSystemHandle.remove() demo](https://filesystemhandle-remove.glitch.me/) (see the [source code](https://glitch.com/edit/#!/filesystemhandle-remove)) is a file creator app. You can enter text into the {{htmlelement("textarea")}} and press the "Save file" {{htmlelement("button")}}, and the app will open a file picker allowing you to save that text onto your local file system in a text file of your choice. You can also choose to delete files you create.
```

Elchi3 reviewed (changes requested):
Just some editorial nits.

In general I don't know what to think about example code that is on glitch and all the explanatory text on the MDN page. It doesn't look like the other File System Access pages do that, so this page seems a little off to the rest and maybe not so integrated. (Why aren't you updating other File API pages to link and advertise for this method if it is so convenient?) 

I'm not familiar enough with the whole of these File API docs to really judge this, though.

chrisdavidmills committed:
Update files/en-us/web/api/filesystemhandle/remove/index.md

Co-authored-by: Florian Scholz <fs@florianscholz.com>

chrisdavidmills committed:
Update files/en-us/web/api/filesystemhandle/remove/index.md

Co-authored-by: Florian Scholz <fs@florianscholz.com>

chrisdavidmills committed:
Update files/en-us/web/api/filesystemhandle/remove/index.md

Co-authored-by: Florian Scholz <fs@florianscholz.com>

chrisdavidmills committed:
Merge branch 'main' into add-filesystemhandle-remove

chrisdavidmills commented:
Thanks @Elchi3. Good comments — I've committed them all. I was impressed by such a curmudgeonly display. You are catching up with @wbamberg ;-)

For the example — yeah, I know what you mean. We should strive for consistency. My thought process was I wanted to write a small demo to make sure I understood how the feature works before documenting it, but then I looked at it and thought it was pretty useful, so I decided to add it to the doc and explain how it works a bit. 

A better place to put this kind of info is probably a "Using..." article, but there wasn't one of those already available to add it to.

And Glitch is such a convenient way to write a demo and link to both it and the source code. The Google folks all use it as standard, so I've been linking to quite a few Glitch demos anyway (most not written by me).

I wonder if a better way would just be to show less tutorial content on the ref page and cover it more in the demo comments...?

wbamberg commented:
> Thanks @Elchi3. Good comments — I've committed them all. I was impressed by such a curmudgeonly display. You are catching up with @wbamberg ;-)
> 
> For the example — yeah, I know what you mean. We should strive for consistency. My thought process was I wanted to write a small demo to make sure I understood how the feature works before documenting it, but then I looked at it and thought it was pretty useful, so I decided to add it to the doc and explain how it works a bit.
> 
> A better place to put this kind of info is probably a "Using..." article, but there wasn't one of those already available to add it to.
> 
> And Glitch is such a convenient way to write a demo and link to both it and the source code. The Google folks all use it as standard, so I've been linking to quite a few Glitch demos anyway (most not written by me).
> 

I visited https://filesystemhandle-remove.glitch.me/ and clicked "Check out the source code." and got https://glitch.com/edit/404.html?projectDomainOrId=filesystemfilehandle-remove. Is it only me that finds Glitch slow and unreliable?

But even if it does work, it's problematic for us, partly as yet another place we keep examples, but also because it's not in our control, so when we want to update it we can't. Just like pages we have that link to random codepens.

Is there a reason this can't be a live sample? Then we don't have two copies of the source code to keep in sync.

chrisdavidmills commented:
> I visited https://filesystemhandle-remove.glitch.me/ and clicked "Check out the source code." and got https://glitch.com/edit/404.html?projectDomainOrId=filesystemfilehandle-remove.

This was because I changed the name of the project, but didn't update the URL that link was pointing to. Fixed now.

> But even if it does work, it's problematic for us, partly as yet another place we keep examples, but also because it's not in our control, so when we want to update it we can't. Just like pages we have that link to random codepens. Also, it means the sample code lives in two places - in the page and in the Glitch instance. So if we, for instance, Prettify the page, it's instantly out of sync with the real code.
> 
> Is there a reason this can't be a live sample?

I think live samples are fine for smaller snippets. But when the example becomes bigger and spans multiple complete files, you then have this situation when you have huge piles of code in the source for the page, and then have to mess about with hiding certain bits of code, and not others, and/or having multiple copies of the code — the truncated bit you show to the reader and the bit that actually powers the live demo. In short, I think that gets just as difficult and annoying to maintain as having it somewhere else. And when you've got a separate example somewhere like Glitch, at least you can easily send just a demo to people, which is what people often want.

The one point I definitely agree with is having control over the code and being able to prettify it, etc. Perhaps to mitigate this problem, I could put bigger examples in one of the MDN example repos instead. https://github.com/mdn/dom-examples maybe? And then use the EmbedGHLiveSample macro to embed them?

wbamberg commented:
> > I visited https://filesystemhandle-remove.glitch.me/ and clicked "Check out the source code." and got https://glitch.com/edit/404.html?projectDomainOrId=filesystemfilehandle-remove.
> 
> This was because I changed the name of the project, but didn't update the URL that link was pointing to. Fixed now.
> 
> > But even if it does work, it's problematic for us, partly as yet another place we keep examples, but also because it's not in our control, so when we want to update it we can't. Just like pages we have that link to random codepens. Also, it means the sample code lives in two places - in the page and in the Glitch instance. So if we, for instance, Prettify the page, it's instantly out of sync with the real code.
> > Is there a reason this can't be a live sample?
> 
> I think live samples are fine for smaller snippets. But when the example becomes bigger and spans multiple complete files, you then have this situation when you have huge piles of code in the source for the page, and then have to mess about with hiding certain bits of code, and not others, and/or having multiple copies of the code — the truncated bit you show to the reader and the bit that actually powers the live demo. In short, I think that gets just as difficult and annoying to maintain as having it somewhere else.

That's fair, but this is quite small - the JS is only 88 lines. And you could have a much smaller example:

```js
let handle = null;

const createBtn = document.querySelector("#create");
createBtn.addEventListener("click", createFile);

async function createFile() {
	handle = await window.showSaveFilePicker();
	const writable = await handle.createWritable();
	await writable.write("hello there");
	await writable.close();
}

const deleteBtn = document.querySelector("#delete");
deleteBtn.addEventListener("click", deleteFile);

function deleteFile() {
	if (handle) {
		handle.remove();
		console.log(`Deleted ${handle.name}`);
	}
}
```

The trick is, that `showSaveFilePicker` doesn't work in cross-origin iframes (which we apparently don't document?) so neither EmbedLiveSample nor EmbedGHLiveSample would work :(.

> And when you've got a separate example somewhere like Glitch, at least you can easily send just a demo to people, which is what people often want.
> 
> The one point I definitely agree with is having control over the code and being able to prettify it, etc. Perhaps to mitigate this problem, I could put bigger examples in one of the MDN example repos instead. https://github.com/mdn/dom-examples maybe? And then use the EmbedGHLiveSample macro to embed them?

I think it's better to use dom-examples than to use Glitch, for sure. When we can, though, I really think live samples are much better for maintenance.

Anyway I'll stop derailing this PR now.

a-sully commented:
> The trick is, that `showSaveFilePicker` doesn't work in cross-origin iframes (which we apparently don't document?) so neither EmbedLiveSample nor EmbedGHLiveSample would work :(.

For this example, you could show use of the `remove()` method from the OPFS, which does not require permission prompts to access (and is supported on all browsers)

The challenge there is showing a user that the file exists at all without using something like [this extension](https://chrome.google.com/webstore/detail/acndjpgkpaclldomagafnognkcgjignd)... or until browsers better support the OPFS in DevTools, though I'm not sure if pointing users to DevTools is common practice for this type of demo either

chrisdavidmills commented:
> Anyway I'll stop derailing this PR now.

No worries; it's good food for thought. I'll look at putting these examples on dom-examples in future if they are big, or as live samples if they are small and will work on live samples.

chrisdavidmills committed:
Merge branch 'main' into add-filesystemhandle-remove

chrisdavidmills commented:
So, can I get this one merged now?

Elchi3 reviewed (approved):
Let's merge this but please don't always default to Glitch going forward.
(Hoping one year we will figure out a decent system for example code, it is not a trivial project).
